src/System.Collections.Immutable/tests/ImmutableArrayTest.cs

```diff
@@ -1129,298 +1642,509 @@ public void ConcatEdgeCases()
+        public static IEnumerable<object[]> IStructuralComparableCompareToNullComparerArgumentInvalidData()
+        {
+            yield return new object[] { null, null };
+            yield return new object[] { null, ImmutableArray.Create(1, 2, 3) };
+            yield return new object[] { new[] { 1, 2, 3 }, null };
+        }
 
-            Assert.Equal(array.CompareTo(equalArray, Comparer<int>.Default), immArray.CompareTo(equalImmArray, Comparer<int>.Default));
+        [Theory]
+        [MemberData(nameof(IStructuralComparableCompareToNullComparerNullReferenceInvalidData))]
+        public void IStructuralComparableCompareToNullComparerNullReferenceInvalid(IEnumerable<int> source, object other)
+        {
+            var array = source.ToImmutableArray();
+            Assert.Throws<NullReferenceException>(() => ((IStructuralComparable)array).CompareTo(other, comparer: null));
 
-            Assert.Throws<ArgumentException>("other", () => array.CompareTo(longerArray, Comparer<int>.Default));
-            Assert.Throws<ArgumentException>("other", () => immArray.CompareTo(longerImmArray, Comparer<int>.Default));
+            if (other == null)
+            {
+                Assert.Throws<NullReferenceException>(() => ((IStructuralComparable)array).CompareTo(s_emptyDefault, comparer: null));
+            }
+        }
 
-            var list = new List<int> { 1, 2, 3 };
-            Assert.Throws<ArgumentException>("other", () => array.CompareTo(list, Comparer<int>.Default));
-            Assert.Throws<ArgumentException>("other", () => immArray.CompareTo(list, Comparer<int>.Default));
+        public static IEnumerable<object[]> IStructuralComparableCompareToNullComparerNullReferenceInvalidData()
+        {
+            // This was not fixed for compatability reasons. See https://github.com/dotnet/corefx/issues/13410
+            yield return new object[] { new[] { 1, 2, 3 }, new[] { 1, 2, 3 } };
+            yield return new object[] { new[] { 1, 2, 3 }, ImmutableArray.Create(1, 2, 3) };
+            // Cache this into a local so the comparands are reference-equal.
+            var oneTwoThree = ImmutableArray.Create(1, 2, 3);
+            yield return new object[] { oneTwoThree, oneTwoThree };
         }
 
-        [Fact]
-        public void StructuralComparableArrayInterop()
+        [Theory]
+        [MemberData(nameof(IStructuralComparableCompareToData))]
+        public void IStructuralComparableCompareTo(IEnumerable<int> source, object other, IComparer comparer, int expected)
         {
-            IStructuralComparable array = new int[3] { 1, 2, 3 };
-            IStructuralComparable equalArray = new int[3] { 1, 2, 3 };
-            IStructuralComparable immArray = ImmutableArray.Create((int[])array);
-            IStructuralComparable equalImmArray = ImmutableArray.Create((int[])equalArray);
+            var array = source?.ToImmutableArray() ?? s_emptyDefault;
```

Comment:
I got a bit lazy here and used `null` for source/other to represent a default immutable array, instead of a `StrongBox`.